Add playthroughs

busy-duck opened the pull request:
chimps play-through for map middle of the road using sauda (also for tinkerton)
please verify it (it worked two times on my laptop) and then run additional script to add it if fitting. 

I am not sure I should officially add it to the readme, etc. Because there are issues when creating the playthrough: I changed the keybinding.json but when creating the playthrough the towers are incorrect. I manually change it but I don't know if it could affect people with other keybindings.

busy-duck commented:
When running the update readme or update plans program I get a message that says nothing has changed. So these new playthroughs are not updated in the readme yet.
Most (if not all I do not exactly remember) also work on impoppable mode with monkey Knowledge. But they should be retested to be sure 😄 !

ANRAR4 commented:
Thanks for recording this many playthroughs

So far i only tested them regarding placement on a resolution of 2560x1440 using `py replay.py validate all -nv -nomk` (ideally all playthroughs should at least work on 1920x1080 and 2560x1440)
In doing so I found 4 maps where the tower placement doesn't work on 2560x1440:
```candy_falls#chimps#1920x1080.btd6	hero placement
enchanted_glade#chimps#1920x1080.btd6	sub0 placement
encrypted#chimps#1920x1080.btd6		hero placement
frozen_over#chimps#1920x1080.btd6	village0 placement
```

As well as some playthroughs with 100% discounted upgrades, which also fail validation if the 100% discount isn't specified (`upgrade bomb0 path 0 with 100% discount`)
```
chutes#chimps#1920x1080.btd6
off_the_coast#chimps#1920x1080.btd6
spice_islands#chimps#1920x1080.btd6
sulfur_springs#chimps#1920x1080.btd6
```

Could you tweak the placements and add ` with 100% discount` to the free upgrades?

And just to make sure: You didn't use any abilities while recording these playthroughs and you also recorded all obstacle removales?

> I am not sure I should officially add it to the readme, etc. Because there are issues when creating the playthrough: I changed the keybinding.json but when creating the playthrough the towers are incorrect. I manually change it but I don't know if it could affect people with other keybindings.

Did this issue persist? As far as i can tell the keybinds shouldn't only be sourced from `keybinds.json`

> When running the update readme or update plans program I get a message that says nothing has changed. So these new playthroughs are not updated in the readme yet.

The table should have been updated, the script output might be somewhat misleading as it is only refering to the other automatically updatable segments.

> Most (if not all I do not exactly remember) also work on impoppable mode with monkey Knowledge. But they should be retested to be sure 😄 !

Having a corresponding table entry would require to copy said playthroughs and specifying impoppable as its native gamemode

Regarding the README.md changes: The former indentation should be correct, and regarding the addition of `up to 3.12.5` I would want to test it further, so could you remove those for now?

Due to not having all chimps runs unlocked I can't test all playthrough, although I will test some

busy-duck commented:
I don't have access to my computer right now but I'll try to update the playthroughs in the week.
 
> And just to make sure: You didn't use any abilities while recording these playthroughs and you also recorded all obstacle removals?

Yes the only obstacle I removed is in encrypted ABR, which I haven't commit yet.

> > I am not sure I should officially add it to the readme, etc. Because there are issues when creating the playthrough: I changed the keybinding.json but when creating the playthrough the towers are incorrect. I manually change it but I don't know if it could affect people with other keybindings.
> Did this issue persist? As far as i can tell the keybinds shouldn't only be sourced from `keybinds.json`

Absolutely, I don't remember for certain but when selecting a monkey with left ctrl it says placing a mortar or other towers (so I end up with 20 or so towers that aren't actually and the upgrade does not correspond to the tower I actually upgraded), when placing a dart it sometimes it says placing a wizard and so on
I ended up manually writing down my actions and then replacing the plythroughs with the correct action. i only use the program to log my tower position.
 
> > When running the update readme or update plans program I get a message that says nothing has changed. So these new playthroughs are not updated in the readme yet. 
> The table should have been updated, the script output might be somewhat misleading as it is only refering to the other automatically updatable segments.

oh ok but I get the same output message with generate_supported_maps_table.py

> >  Most (if not all I do not exactly remember) also work on impoppable mode with monkey Knowledge. But they should be retested to be sure 😄 !
> Having a corresponding table entry would require to copy said playthroughs and specifying impoppable as its native gamemode

Does this mean copying the file and changing its name with impoppable ?

> Regarding the README.md changes: The former indentation should be correct, and regarding the addition of `up to 3.12.5` I would want to test it further, so could you remove those for now?

I can do that but the indentation level does not seem correct: the "selecting a monkey" title is at the same level as the action that would require a monkey to be selected as well as action that do not require a monkey to be selected (here: "Awaiting a round" and "Monkey naming").

> Due to not having all chimps runs unlocked I can't test all playthrough, although I will test some

don't worry although there is only one playthrough I haven't tested: haunted chimps.
I don't think it will work so I'll either work on it or delete it completely.

busy-duck commented:
Hi I have a question regarding the tower placement issues, I tested the three maps and they are validated, should I place the towers a little to the left or to the right? I don't really understand the issue.

For my keybind.json I will copy it in a while I need to access my gaming computer. But I should tell you that my keyboard layout is 'french-pc', however the btd6 hotkeys are the original ones. As in the dart monkey is displayed as the hotkey "q" but I actually have to press the key "a" in order to get a dart monkey. (I think the hotkey displayed in the game is the "position" of the key as opposed to the actual input of the key...).

In regards to the readme titles: when first reading it I didn't understand what the actions that required selecting a monkey were. I feel like indenting these actions make more sense. The same way that 'Placing a monkey' and 'Removing an obstacle' are sub-levels of 'Recording playthroughs'. In the same logic, the actions that required selecting a monkey would be sub-levels of 'selecting a monkey' while still being in the 'Recording playthroughs' section.

busy-duck commented:
Here's the keybinds.json
{
    "monkeys": {
        "dart": "a",
        "boomerang": "z",
        "bomb": "e",
        "tack": "r",
        "ice": "t",
        "glue": "y",
        "sniper": "w",
        "sub": "x",
        "buccaneer": "c",
        "ace": "v",
        "heli": "b",
        "mortar": "n",
        "dartling": ",",
        "wizard": "q",
        "super": "s",
        "ninja": "d",
        "alchemist": "f",
        "druid": "g",
        "farm": "h",
        "engineer": "l",
        "spike": "j",
        "village": "k",
        "beasthandler": "i",
        "mermonkey": "o",
        "desperado": "p",
        "hero": "u"
    },
    "path": {
        "0": 51,
        "1": 52,
        "2": 53
    },
    "others": {
        "retarget": "{Tab}",
        "special": "{Up}",
        "sell": "{Backspace}",
        "play": "{Space}"
    },
    "recording": {
        "select_monkey": "ctrl",
        "remove_obstacle": "right shift",
        "retarget": "tab",
        "sell": "backspace",
        "monkey_special": "up",
        "await_round": "left shift"
    }
}

I should note that the replay process works fine. Issues only arise when recording a playthrough.
